<!DOCTYPE html>
<html lang="en" dir="auto" data-theme="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">
<title>Realising the full potential of MR-PHeWAS in cancer | Juiceright</title>
<meta name="keywords" content="">
<meta name="description" content="笔记Realising the full potential of MR-PHeWAS in cancer.

️ 出版年份: 2021\  出版期刊: British journal of cancer\  影响因子: 8.8\  JCR分区: 1\  DOI:10.1038/s41416-020-01165-0

 文章作者: Bowden Jack

 摘要: MR-PHeWAS is a powerful new design for discovering causal mechanisms between a disease and its many candidate risk factors in a hypothesis-free manner. This technique has great potential in the field of cancer research, provided that both powerful and principled statistical approaches are used.

 结论:
“This technique, termed ‘MR-PheWAS’ (MR-phenomewide association study), is used to prioritise further epidemiological studies and has also been used to prioritise potential drug targets in the pharmaceutical arena.4” (Bowden, 2021, p. 1)">
<meta name="author" content="">
<link rel="canonical" href="https://cywd123.github.io/old/posts/%E7%AC%94%E8%AE%B0realising-the-full-potential-of-mr-phewas-in-cancer/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.343cc480b9ffc8f04ccbe5e968ad674880cab773ec19905e93033065c1e7a804.css" integrity="sha256-NDzEgLn/yPBMy&#43;XpaK1nSIDKt3PsGZBekwMwZcHnqAQ=" rel="preload stylesheet" as="style">
<link rel="icon" href="https://cywd123.github.io/favicon.ico">
<link rel="icon" type="image/png" sizes="16x16" href="https://cywd123.github.io/favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="https://cywd123.github.io/favicon-32x32.png">
<link rel="apple-touch-icon" href="https://cywd123.github.io/apple-touch-icon.png">
<link rel="mask-icon" href="https://cywd123.github.io/safari-pinned-tab.svg">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://cywd123.github.io/old/posts/%E7%AC%94%E8%AE%B0realising-the-full-potential-of-mr-phewas-in-cancer/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --code-block-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
                color-scheme: dark;
            }

            .list {
                background: var(--theme);
            }

            .toc {
                background: var(--entry);
            }
        }

        @media (prefers-color-scheme: light) {
            .list::-webkit-scrollbar-thumb {
                border-color: var(--code-bg);
            }
        }

    </style>
</noscript>
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.querySelector("html").dataset.theme = 'dark';
    } else if (localStorage.getItem("pref-theme") === "light") {
       document.querySelector("html").dataset.theme = 'light';
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.querySelector("html").dataset.theme = 'dark';
    } else {
        document.querySelector("html").dataset.theme = 'light';
    }

</script><meta property="og:url" content="https://cywd123.github.io/old/posts/%E7%AC%94%E8%AE%B0realising-the-full-potential-of-mr-phewas-in-cancer/">
  <meta property="og:site_name" content="Juiceright">
  <meta property="og:title" content="Realising the full potential of MR-PHeWAS in cancer">
  <meta property="og:description" content="笔记Realising the full potential of MR-PHeWAS in cancer. ️ 出版年份: 2021\ 出版期刊: British journal of cancer\ 影响因子: 8.8\ JCR分区: 1\ DOI:10.1038/s41416-020-01165-0
文章作者: Bowden Jack
摘要: MR-PHeWAS is a powerful new design for discovering causal mechanisms between a disease and its many candidate risk factors in a hypothesis-free manner. This technique has great potential in the field of cancer research, provided that both powerful and principled statistical approaches are used. 结论: “This technique, termed ‘MR-PheWAS’ (MR-phenomewide association study), is used to prioritise further epidemiological studies and has also been used to prioritise potential drug targets in the pharmaceutical arena.4” (Bowden, 2021, p. 1)">
  <meta property="og:locale" content="en-us">
  <meta property="og:type" content="article">
    <meta property="article:section" content="old">
    <meta property="article:published_time" content="2024-01-20T11:28:52+00:00">
    <meta property="article:modified_time" content="2025-12-01T09:31:17+00:00">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Realising the full potential of MR-PHeWAS in cancer">
<meta name="twitter:description" content="笔记Realising the full potential of MR-PHeWAS in cancer.

️ 出版年份: 2021\  出版期刊: British journal of cancer\  影响因子: 8.8\  JCR分区: 1\  DOI:10.1038/s41416-020-01165-0

 文章作者: Bowden Jack

 摘要: MR-PHeWAS is a powerful new design for discovering causal mechanisms between a disease and its many candidate risk factors in a hypothesis-free manner. This technique has great potential in the field of cancer research, provided that both powerful and principled statistical approaches are used.

 结论:
“This technique, termed ‘MR-PheWAS’ (MR-phenomewide association study), is used to prioritise further epidemiological studies and has also been used to prioritise potential drug targets in the pharmaceutical arena.4” (Bowden, 2021, p. 1)">


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "Olds",
      "item": "https://cywd123.github.io/old/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "Realising the full potential of MR-PHeWAS in cancer",
      "item": "https://cywd123.github.io/old/posts/%E7%AC%94%E8%AE%B0realising-the-full-potential-of-mr-phewas-in-cancer/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "Realising the full potential of MR-PHeWAS in cancer",
  "name": "Realising the full potential of MR-PHeWAS in cancer",
  "description": "笔记Realising the full potential of MR-PHeWAS in cancer. ️ 出版年份: 2021\\ 出版期刊: British journal of cancer\\ 影响因子: 8.8\\ JCR分区: 1\\ DOI:10.1038/s41416-020-01165-0\n文章作者: Bowden Jack\n摘要: MR-PHeWAS is a powerful new design for discovering causal mechanisms between a disease and its many candidate risk factors in a hypothesis-free manner. This technique has great potential in the field of cancer research, provided that both powerful and principled statistical approaches are used. 结论: “This technique, termed ‘MR-PheWAS’ (MR-phenomewide association study), is used to prioritise further epidemiological studies and has also been used to prioritise potential drug targets in the pharmaceutical arena.4” (Bowden, 2021, p. 1)\n",
  "keywords": [
    
  ],
  "articleBody": "笔记Realising the full potential of MR-PHeWAS in cancer. ️ 出版年份: 2021\\ 出版期刊: British journal of cancer\\ 影响因子: 8.8\\ JCR分区: 1\\ DOI:10.1038/s41416-020-01165-0\n文章作者: Bowden Jack\n摘要: MR-PHeWAS is a powerful new design for discovering causal mechanisms between a disease and its many candidate risk factors in a hypothesis-free manner. This technique has great potential in the field of cancer research, provided that both powerful and principled statistical approaches are used. 结论: “This technique, termed ‘MR-PheWAS’ (MR-phenomewide association study), is used to prioritise further epidemiological studies and has also been used to prioritise potential drug targets in the pharmaceutical arena.4” (Bowden, 2021, p. 1)\n“novel methods are urgently needed in order to fully exploit the MR-PHeWAS design.” (Bowden, 2021, p. 2)\n原文提出MR-PHeWAS的局限性： 由于分析次数多, Bonferroni检验过于严格导致没有阳性结果？？\n“In their analysis, no single phenotype was estimated to have a strong enough causal effect to fall below a pre-specified Bonferroni-adjusted 5% type I error threshold, although 13 phenotypes showed suggestive evidence of a causal association (P \u003c 5%). Furthermore, when fully pleiotropy-robust MR methods were used, only telomere length, low-density lipoprotein and glycated haemoglobin remained suggestive. Although the analysis was inconclusive, and the methodology used sound, their work highlights a number of current limitations in the statistical methods routinely applied to MR-PHeWAS.” (Bowden, 2021, p. 1)\n多数表型都是由单一的SNP驱动，所以存在异质性？？\n“Many of the strongest results in their analysis were driven by a single variant (e.g. the effect of telomere length on glioma risk through the TERT gene), but the precision of their overall causal effect was dramatically diminished due to the presence of substantial heterogeneity across the remaining SNPs. Modern MR approaches generally utilise large numbers of SNPs, but interpret heterogeneity in causal estimates across SNPs as a sign of horizontal pleiotropy.3” (Bowden, 2021, p. 1)\n3. 如何从大量的有bias的数据中分离出叫小部分的可靠数据？？\n“However, they still assume that the majority of the genetic signals are correct, or they are all correct ‘on average’. This may be true, but it could also be the case, for example, that TERT is the only reliable genetic instrument in the analysis. This presents a future challenge for MR approaches in being able to separate out the small kernel of truth in a larger sea of biased data.” (Bowden, 2021, p. 1)\n总结 我的想法: Bonferroni检验过于严格，那就试试更加宽松的多重检验方法 比如FDR\n单一SNP也能检验异质性，用SMR的HEIDI方法，不过HEIDI方法更多用于MR药靶（QTL作为表型）。\n可以试试三样本孟德尔随机化，三样本孟德尔随机化能够在许多弱工具变量中建立因果关系（不必强求IV的F值大于10）\n(“三样本孟德尔随机化简介-腾讯云开发者社区-腾讯云”)\n",
  "wordCount" : "369",
  "inLanguage": "en",
  "datePublished": "2024-01-20T11:28:52Z",
  "dateModified": "2025-12-01T09:31:17.685Z",
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://cywd123.github.io/old/posts/%E7%AC%94%E8%AE%B0realising-the-full-potential-of-mr-phewas-in-cancer/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Juiceright",
    "logo": {
      "@type": "ImageObject",
      "url": "https://cywd123.github.io/favicon.ico"
    }
  }
}
</script>
</head>
<body id="top">
    <header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://cywd123.github.io/" accesskey="h" title="Juiceright (Alt + H)">Juiceright</a>
            <div class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)" aria-label="Toggle theme">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="18" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="18" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
            </div>
        </div>
        <ul id="menu">
            <li>
                <a href="https://cywd123.github.io/search/" title="Search (Alt &#43; /)" accesskey=/>
                    <span>Search</span>
                </a>
            </li>
            <li>
                <a href="https://cywd123.github.io/tags/" title="Tags">
                    <span>Tags</span>
                </a>
            </li>
            <li>
                <a href="https://cywd123.github.io/archives/" title="Archives">
                    <span>Archives</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    
    <h1 class="post-title entry-hint-parent">
      Realising the full potential of MR-PHeWAS in cancer
    </h1>
    <div class="post-meta"><span title='2024-01-20 11:28:52 +0000 UTC'>January 20, 2024</span>&nbsp;·&nbsp;<span>2 min</span>

</div>
  </header> 
  <div class="post-content"><h1 id="realising-the-full-potential-of-mr-phewas-in-cancer"><strong><span style="color: #7CB342">笔记</span></strong><span style="color: #7CB342">Realising the full potential of MR-PHeWAS in cancer.</span><a hidden class="anchor" aria-hidden="true" href="#realising-the-full-potential-of-mr-phewas-in-cancer">#</a></h1>
<hr>
<p><strong><span style="color: #1976D2">️ 出版年份: </span>2021\ <span style="color: #1976D2"> 出版期刊: </span>British journal of cancer\ <span style="color: #1976D2"> 影响因子: </span>8.8\ <span style="color: #1976D2"> JCR分区: </span>1\ <span style="color: #1976D2"> DOI:10.1038/s41416-020-01165-0<br></span></strong></p>
<hr>
<p><span style="color: #1976D2"> 文章作者: </span>Bowden Jack</p>
<hr>
<h2 id="摘要-mr-phewas-is-a-powerful-new-design-for-discovering-causal-mechanisms-between-a-disease-and-its-many-candidate-risk-factors-in-a-hypothesis-free-manner-this-technique-has-great-potential-in-the-field-of-cancer-research-provided-that-both-powerful-and-principled-statistical-approaches-are-used"><span style="color: #F06292"> 摘要: </span>MR-PHeWAS is a powerful new design for discovering causal mechanisms between a disease and its many candidate risk factors in a hypothesis-free manner. This technique has great potential in the field of cancer research, provided that both powerful and principled statistical approaches are used.<a hidden class="anchor" aria-hidden="true" href="#摘要-mr-phewas-is-a-powerful-new-design-for-discovering-causal-mechanisms-between-a-disease-and-its-many-candidate-risk-factors-in-a-hypothesis-free-manner-this-technique-has-great-potential-in-the-field-of-cancer-research-provided-that-both-powerful-and-principled-statistical-approaches-are-used">#</a></h2>
<hr>
<h2 id="结论"><span style="background-color: #ce93d8"> 结论:</span><a hidden class="anchor" aria-hidden="true" href="#结论">#</a></h2>
<p>“This technique, termed ‘MR-PheWAS’ (MR-phenomewide association study), is used to prioritise further epidemiological studies and has also been used to prioritise potential drug targets in the pharmaceutical arena.4” <span class="citation" data-citation="%7B%22citationItems%22%3A%5B%7B%22uris%22%3A%5B%22http%3A%2F%2Fzotero.org%2Fusers%2F13422775%2Fitems%2F5Y35DM5R%22%5D%2C%22locator%22%3A%221%22%7D%5D%2C%22properties%22%3A%7B%7D%7D" ztype="zcitation">(<span class="citation-item"><a href="zotero://select/library/items/5Y35DM5R">Bowden, 2021, p. 1</a></span>)</span></p>
<p><span class="highlight" data-annotation="%7B%22attachmentURI%22%3A%22http%3A%2F%2Fzotero.org%2Fusers%2F13422775%2Fitems%2FEA5ZVH7J%22%2C%22pageLabel%22%3A%222%22%2C%22position%22%3A%7B%22pageIndex%22%3A1%2C%22rects%22%3A%5B%5B105.676%2C604.176%2C293.414%2C612.416%5D%2C%5B45.354%2C594.198%2C184.389%2C602.438%5D%5D%7D%2C%22citationItem%22%3A%7B%22uris%22%3A%5B%22http%3A%2F%2Fzotero.org%2Fusers%2F13422775%2Fitems%2F5Y35DM5R%22%5D%2C%22locator%22%3A%222%22%7D%7D" ztype="zhighlight"><a href="zotero://open-pdf/library/items/EA5ZVH7J?page=2">“novel methods are urgently needed in order to fully exploit the MR-PHeWAS design.”</a></span> <span class="citation" data-citation="%7B%22citationItems%22%3A%5B%7B%22uris%22%3A%5B%22http%3A%2F%2Fzotero.org%2Fusers%2F13422775%2Fitems%2F5Y35DM5R%22%5D%2C%22locator%22%3A%222%22%7D%5D%2C%22properties%22%3A%7B%7D%7D" ztype="zcitation">(<span class="citation-item"><a href="zotero://select/library/items/5Y35DM5R">Bowden, 2021, p. 2</a></span>)</span></p>
<hr>
<h2 id="原文提出mr-phewas的局限性"><span style="background-color: #ffd40080">原文提出MR-PHeWAS的局限性：</span><a hidden class="anchor" aria-hidden="true" href="#原文提出mr-phewas的局限性">#</a></h2>
<p>由于分析次数多, Bonferroni检验过于严格导致没有阳性结果？？</p>
<p>“In their analysis, no single phenotype was estimated to have a strong enough causal effect to fall below a pre-specified Bonferroni-adjusted 5% type I error threshold, although 13 phenotypes showed suggestive evidence of a causal association (P &lt; 5%). Furthermore, when fully pleiotropy-robust MR methods were used, only telomere length, low-density lipoprotein and glycated haemoglobin remained suggestive. Although the analysis was inconclusive, and the methodology used sound, their work highlights a number of current limitations in the statistical methods routinely applied to MR-PHeWAS.” <span class="citation" data-citation="%7B%22citationItems%22%3A%5B%7B%22uris%22%3A%5B%22http%3A%2F%2Fzotero.org%2Fusers%2F13422775%2Fitems%2F5Y35DM5R%22%5D%2C%22locator%22%3A%221%22%7D%5D%2C%22properties%22%3A%7B%7D%7D" ztype="zcitation">(<span class="citation-item"><a href="zotero://select/library/items/5Y35DM5R">Bowden, 2021, p. 1</a></span>)</span></p>
<p>多数表型都是由单一的SNP驱动，所以存在异质性？？</p>
<p>“Many of the strongest results in their analysis were driven by a single variant (e.g. the effect of telomere length on glioma risk through the TERT gene), but the precision of their overall causal effect was dramatically diminished due to the presence of substantial heterogeneity across the remaining SNPs. Modern MR approaches generally utilise large numbers of SNPs, but interpret heterogeneity in causal estimates across SNPs as a sign of horizontal pleiotropy.3” <span class="citation" data-citation="%7B%22citationItems%22%3A%5B%7B%22uris%22%3A%5B%22http%3A%2F%2Fzotero.org%2Fusers%2F13422775%2Fitems%2F5Y35DM5R%22%5D%2C%22locator%22%3A%221%22%7D%5D%2C%22properties%22%3A%7B%7D%7D" ztype="zcitation">(<span class="citation-item"><a href="zotero://select/library/items/5Y35DM5R">Bowden, 2021, p. 1</a></span>)</span></p>
<p>3. 如何从大量的有bias的数据中分离出叫小部分的可靠数据？？</p>
<p>“However, they still assume that the majority of the genetic signals are correct, or they are all correct ‘on average’. This may be true, but it could also be the case, for example, that TERT is the only reliable genetic instrument in the analysis. This presents a future challenge for MR approaches in being able to separate out the small kernel of truth in a larger sea of biased data.” <span class="citation" data-citation="%7B%22citationItems%22%3A%5B%7B%22uris%22%3A%5B%22http%3A%2F%2Fzotero.org%2Fusers%2F13422775%2Fitems%2F5Y35DM5R%22%5D%2C%22locator%22%3A%221%22%7D%5D%2C%22properties%22%3A%7B%7D%7D" ztype="zcitation">(<span class="citation-item"><a href="zotero://select/library/items/5Y35DM5R">Bowden, 2021, p. 1</a></span>)</span></p>
<hr>
<h2 id="heading"><span style="color: #F06292"> </span><strong><span style="color: #F06292">总结</span></strong><a hidden class="anchor" aria-hidden="true" href="#heading">#</a></h2>
<h3 id="我的想法"><span style="background-color: #9ccc65">   我的想法:</span><a hidden class="anchor" aria-hidden="true" href="#我的想法">#</a></h3>
<ol>
<li>
<p>Bonferroni检验过于严格，那就试试更加宽松的多重检验方法 比如FDR</p>
</li>
<li>
<p>单一SNP也能检验异质性，用SMR的HEIDI方法，不过HEIDI方法更多用于MR药靶（QTL作为表型）。</p>
</li>
<li>
<p>可以试试三样本孟德尔随机化，三样本孟德尔随机化能够在许多弱工具变量中建立因果关系（不必强求IV的F值大于10）</p>
<p><span class="citation" data-citation="%7B%22citationItems%22%3A%5B%7B%22uris%22%3A%5B%22http%3A%2F%2Fzotero.org%2Fusers%2F13422775%2Fitems%2FBUE9X2QK%22%5D%7D%5D%2C%22properties%22%3A%7B%7D%7D" ztype="zcitation">(<span class="citation-item"><a href="zotero://select/library/items/BUE9X2QK">“三样本孟德尔随机化简介-腾讯云开发者社区-腾讯云”</a></span>)</span></p>
</li>
</ol>
<hr>


  </div>

  <footer class="post-footer">
    <ul class="post-tags">
    </ul>
  </footer>
</article>
    </main>
    

<footer class="footer">
        <span>&copy; 2025 <a href="https://cywd123.github.io/">Juiceright</a></span> · 

    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://github.com/adityatelange/hugo-PaperMod/" rel="noopener" target="_blank">PaperMod</a>
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu');
    if (menu) {
        
        const scrollPosition = localStorage.getItem("menu-scroll-position");
        if (scrollPosition) {
            menu.scrollLeft = parseInt(scrollPosition, 10);
        }
        
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        const html = document.querySelector("html");
        if (html.dataset.theme === "dark") {
            html.dataset.theme = 'light';
            localStorage.setItem("pref-theme", 'light');
        } else {
            html.dataset.theme = 'dark';
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerHTML = 'copy';

        function copyingDone() {
            copybutton.innerHTML = 'copied!';
            setTimeout(() => {
                copybutton.innerHTML = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>
</body>

</html>
