

<!DOCTYPE html>
<html lang="en" itemscope itemtype="http://schema.org/WebPage">
  <head>
    

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0">

 


  <script type="application/ld+json">
{
    "@context": "http://schema.org",
    "@type": "WebSite",
    "name": "Juiceright",
    
    "url": "https:\/\/cywd123.github.io\/"
}
</script><script type="application/ld+json">
{
  "@context": "http://schema.org",
  "@type": "Organization",
  "name": "",
  "url": "https:\/\/cywd123.github.io\/"
  
  
  
  
}
</script>

<meta property="og:title" content="Juiceright" />
<meta property="og:url" content="https://cywd123.github.io/" />
<meta property="og:type" content="website" />
<meta property="og:site_name" content="Juiceright" />

  <meta name="twitter:title" content="Juiceright" />
  <meta name="twitter:card" content="summary_large_image" />
  <meta name="generator" content="Hugo 0.152.2">
  <link rel="alternate" href="https://cywd123.github.io/index.xml" type="application/rss+xml" title="Juiceright"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.22/dist/katex.min.css" integrity="sha384-5TcZemv2l/9On385z///+d7MSYlvIEw9FuZTIdZ14vJLqWphw7e7ZPuOiCHJcFCP" crossorigin="anonymous">
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v7.0.1/css/brands.css" integrity="sha384-+eXGm6TmVoCZR1gX03US+L++L2mJIfOvp2X0+luTuEktXaF5F+e3zn0sO0HBvOyT" crossorigin="anonymous">
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v7.0.1/css/fontawesome.css" integrity="sha384-B5nyPZa7e84uCJEYHtevCdVYtQosWsbDSRtq3cy/jeSVtQW05GL8CKXO7waXsjPK" crossorigin="anonymous">
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v7.0.1/css/regular.css" integrity="sha384-V71z/A/p/DGaZAryHhiSXRQj8HdQmK2nBgL2bX23MGD+h6ZKpyxOO76f5eSXmIaj" crossorigin="anonymous">
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v7.0.1/css/solid.css" integrity="sha384-1IIShlrK5iYgvS7OAgIH3prz4RAFP7hHZ2MaLIb9l3+trFwwnw/NdCWcUHlymZDU" crossorigin="anonymous">
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v7.0.1/css/svg.css" integrity="sha384-iBa7uugTOHXqQ2ngvDYJyKrPPQbd9n2P3EQe+o/0+HZe5TNvV+LO5X0lLLUrXLHP" crossorigin="anonymous">
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v7.0.1/css/svg-with-js.css" integrity="sha384-STWnS5AGz6M2YxsnGAd536EqcE2y+ktGovMMuhtXLsCBtl7lujeQj2eN4eDxa57x" crossorigin="anonymous">
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v7.0.1/css/v4-font-face.css" integrity="sha384-fH83IkgQi7kturzlxpXfNdsR8aUOcqj5HUORRVnSvJgObZkIw6aqKAaQx20IWyB/" crossorigin="anonymous">
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v7.0.1/css/v4-shims.css" integrity="sha384-cCODJHSivNBsaHei/8LC0HUD58kToSbDU+xT7Rs51BO1v/IvgT/uM0W6xMoUqKfn" crossorigin="anonymous">
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v7.0.1/css/v5-font-face.css" integrity="sha384-atSdF89mIn15dv+o2LKkDM6mCh3hwh6A7mwZ6F2it/kenqZbMu+QW/ye1OG4XSWV" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@3.4.1/dist/css/bootstrap.min.css" integrity="sha384-HSMxcRTRxnN+Bdg0JdbxYKrThecOKuH5zCYotlSAcp1+c8xmyTe9GYg1l9a69psu" crossorigin="anonymous"><link rel="stylesheet" href="https://cywd123.github.io/css/main.css" /><link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lora:400,700,400italic,700italic" />
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:300italic,400italic,600italic,700italic,800italic,400,300,600,700,800" /><link rel="stylesheet" href="https://cywd123.github.io/css/syntax.css" /><link rel="stylesheet" href="https://cywd123.github.io/css/codeblock.css" /><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/photoswipe/4.1.2/photoswipe.min.css" integrity="sha384-h/L2W9KefUClHWaty3SLE5F/qvc4djlyR4qY3NUV5HGQBBW7stbcfff1+I/vmsHh" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/photoswipe/4.1.2/default-skin/default-skin.min.css" integrity="sha384-iD0dNku6PYSIQLyfTOpB06F2KCZJAKLOThS5HRe8b3ibhdEQ6eKsFf/EeFxdOt5R" crossorigin="anonymous">

  </head>
  <body>
    <nav class="navbar navbar-default navbar-fixed-top navbar-custom">
  <div class="container-fluid">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#main-navbar">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="https://cywd123.github.io/">Juiceright</a>
    </div>

    <div class="collapse navbar-collapse" id="main-navbar">
      <ul class="nav navbar-nav navbar-right">
        

        

        
      </ul>
    </div>

    

  </div>
</nav>




    


<div class="pswp" tabindex="-1" role="dialog" aria-hidden="true">

<div class="pswp__bg"></div>

<div class="pswp__scroll-wrap">
    
    <div class="pswp__container">
      <div class="pswp__item"></div>
      <div class="pswp__item"></div>
      <div class="pswp__item"></div>
    </div>
    
    <div class="pswp__ui pswp__ui--hidden">
    <div class="pswp__top-bar">
      
      <div class="pswp__counter"></div>
      <button class="pswp__button pswp__button--close" title="Close (Esc)"></button>
      <button class="pswp__button pswp__button--share" title="Share"></button>
      <button class="pswp__button pswp__button--fs" title="Toggle fullscreen"></button>
      <button class="pswp__button pswp__button--zoom" title="Zoom in/out"></button>
      
      
      <div class="pswp__preloader">
        <div class="pswp__preloader__icn">
          <div class="pswp__preloader__cut">
            <div class="pswp__preloader__donut"></div>
          </div>
        </div>
      </div>
    </div>
    <div class="pswp__share-modal pswp__share-modal--hidden pswp__single-tap">
      <div class="pswp__share-tooltip"></div>
    </div>
    <button class="pswp__button pswp__button--arrow--left" title="Previous (arrow left)">
    </button>
    <button class="pswp__button pswp__button--arrow--right" title="Next (arrow right)">
    </button>
    <div class="pswp__caption">
      <div class="pswp__caption__center"></div>
    </div>
    </div>
    </div>
</div>


  
  
  






  

  <header class="header-section ">
    
    
    <div class="intro-header no-img">
      <div class="container">
        <div class="row">
          <div class="col-lg-8 col-lg-offset-2 col-md-10 col-md-offset-1">
            <div class="page-heading">
              
                <h1>Juiceright</h1>
              
              
                <hr class="small">
              
              
              
            </div>
          </div>
        </div>
      </div>
    </div>
  
  </header>


    
  <div role="main" class="container">
    <div class="row">
      <div class="col-lg-8 col-lg-offset-2 col-md-10 col-md-offset-1">
        

        <div class="posts-list">
          
          
            <article class="post-preview">
    <a href="https://cywd123.github.io/posts/go%E5%92%8Ckegg%E5%88%86%E6%9E%90/">
        <h2 class="post-title">GO和KEGG分析</h2>
        
        
        
    </a>

    <p class="post-meta">
        <span class="post-meta">
  
  
  <i class="fas fa-calendar"></i>&nbsp;Posted on August 6, 2023
  
    &nbsp;(Last modified on December 1, 2025)
  
  
  
  
    
      &nbsp;|&nbsp;<i class="fas fa-user"></i>&nbsp;
    
  
  
</span>


    </p>
    <div class="post-entry">
        
        <h1 id="go和kegg分析">GO和KEGG分析</h1>
<p>功能富集分析对于解释转录组学数据至关重要。转录组鉴定了<strong>差异表达基因</strong>后，通常会进行GO或KEGG富集分析，识别这些差异基因的功能或参与调控的通路，来说明<strong>关键基因表达上调/或下调后可能会导致哪功能或通路被激活/或抑制</strong>，进而与表型进行联系。</p>
<p>(其实就是用DEG结果的进行的分析)</p>
<h1 id="为什么要做go和kegg分析">为什么要做GO和KEGG分析</h1>
<p>经过差异表达分析，我们得到了在对照组与实验组中差异表达的基因，说明改变的条件对这些基因的表达产生了影响，但是这样还不够，我们希望进一步知道具体是<strong>对哪些生物学功能/通路产生了影响</strong>，于是需要进行GO分析。</p>
<h1 id="gokegg结果解读">GO、KEGG结果解读</h1>
<p><a href="https://zhuanlan.zhihu.com/p/581773111">富集分析，看完这篇就够</a></p>
<p><a href="https://zhuanlan.zhihu.com/p/339046340">别搜啦！关于富集分析你想知道的这里都有！</a></p>
<p><a href="https://zhuanlan.zhihu.com/p/561522453">使用R包clusterProfiler进行GO/KEGG富集分析（有参/无参）</a></p>
<h1 id="输入数据格式">输入数据格式</h1>
<p>一个表格，分别是Gene_Symbol和logFC，如下图所示：</p>
<p><img src="/images/1741977478264.jpg" alt=""></p>
<h1 id="主要结果图">主要结果图</h1>
<p><img src="/images/1741977482727.jpg" alt=""></p>
<hr>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#a6e22e">library</span>(<span style="color:#e6db74">&#34;clusterProfiler&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">library</span>(<span style="color:#e6db74">&#34;enrichplot&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">library</span>(<span style="color:#e6db74">&#34;ggplot2&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">library</span>(<span style="color:#e6db74">&#34;org.Hs.eg.db&#34;</span>) 
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">library</span>(<span style="color:#e6db74">&#34;GOplot&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#75715e">#转换ID</span>
</span></span><span style="display:flex;"><span>rt<span style="color:#f92672">=</span><span style="color:#a6e22e">read.table</span>(<span style="color:#e6db74">&#34;input.txt&#34;</span>,sep<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;\t&#34;</span>,check.names<span style="color:#f92672">=</span>F,header<span style="color:#f92672">=</span>T)
</span></span><span style="display:flex;"><span>genes<span style="color:#f92672">=</span><span style="color:#a6e22e">as.vector</span>(rt[,<span style="color:#ae81ff">1</span>])
</span></span><span style="display:flex;"><span>entrezIDs <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">mget</span>(genes, org.Hs.egSYMBOL2EG, ifnotfound<span style="color:#f92672">=</span><span style="color:#66d9ef">NA</span>)    <span style="color:#75715e">#找出基因对应的id</span>
</span></span><span style="display:flex;"><span>entrezIDs <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">as.character</span>(entrezIDs)
</span></span><span style="display:flex;"><span>out<span style="color:#f92672">=</span><span style="color:#a6e22e">cbind</span>(rt,entrezID<span style="color:#f92672">=</span>entrezIDs)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">write.table</span>(out,file<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;GO-id.txt&#34;</span>,sep<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;\t&#34;</span>,quote<span style="color:#f92672">=</span>F,row.names<span style="color:#f92672">=</span>F)    <span style="color:#75715e">#输出结果</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#75715e">####GO分析####</span>
</span></span><span style="display:flex;"><span>rt<span style="color:#f92672">=</span><span style="color:#a6e22e">read.table</span>(<span style="color:#e6db74">&#34;GO-id.txt&#34;</span>,sep<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;\t&#34;</span>,header<span style="color:#f92672">=</span>T,check.names<span style="color:#f92672">=</span>F)           
</span></span><span style="display:flex;"><span>rt<span style="color:#f92672">=</span>rt<span style="color:#a6e22e">[is.na</span>(rt[,<span style="color:#e6db74">&#34;entrezID&#34;</span>])<span style="color:#f92672">==</span>F,]                                
</span></span><span style="display:flex;"><span>gene<span style="color:#f92672">=</span>rt<span style="color:#f92672">$</span>entrezID
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#75715e">#GO富集分析</span>
</span></span><span style="display:flex;"><span>kk <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">enrichGO</span>(gene <span style="color:#f92672">=</span> gene,OrgDb <span style="color:#f92672">=</span> org.Hs.eg.db, pvalueCutoff <span style="color:#f92672">=</span><span style="color:#ae81ff">0.05</span>, qvalueCutoff <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.05</span>,ont<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;all&#34;</span>,readable <span style="color:#f92672">=</span>T)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">write.table</span>(kk,file<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;GO.txt&#34;</span>,sep<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;\t&#34;</span>,quote<span style="color:#f92672">=</span>F,row.names <span style="color:#f92672">=</span> F) <span style="color:#75715e">#</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#75715e">#可视化</span>
</span></span><span style="display:flex;"><span><span style="color:#75715e">##条形图</span>
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">pdf</span>(file<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;GO-barplot.pdf&#34;</span>,width <span style="color:#f92672">=</span> <span style="color:#ae81ff">10</span>,height <span style="color:#f92672">=</span> <span style="color:#ae81ff">15</span>)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">barplot</span>(kk, drop <span style="color:#f92672">=</span> <span style="color:#66d9ef">TRUE</span>, showCategory <span style="color:#f92672">=</span><span style="color:#ae81ff">10</span>,label_format<span style="color:#f92672">=</span><span style="color:#ae81ff">100</span>,split<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;ONTOLOGY&#34;</span>) <span style="color:#f92672">+</span> <span style="color:#a6e22e">facet_grid</span>(ONTOLOGY<span style="color:#f92672">~</span>., scale<span style="color:#f92672">=</span><span style="color:#e6db74">&#39;free&#39;</span>)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">dev.off</span>()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#75715e">##气泡图</span>
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">pdf</span>(file<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;GO-bubble.pdf&#34;</span>,width <span style="color:#f92672">=</span> <span style="color:#ae81ff">10</span>,height <span style="color:#f92672">=</span> <span style="color:#ae81ff">15</span>)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">dotplot</span>(kk,showCategory <span style="color:#f92672">=</span> <span style="color:#ae81ff">10</span>,label_format<span style="color:#f92672">=</span><span style="color:#ae81ff">100</span>,split<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;ONTOLOGY&#34;</span>) <span style="color:#f92672">+</span> <span style="color:#a6e22e">facet_grid</span>(ONTOLOGY<span style="color:#f92672">~</span>., scale<span style="color:#f92672">=</span><span style="color:#e6db74">&#39;free&#39;</span>)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">dev.off</span>()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#75715e">##圈图</span>
</span></span><span style="display:flex;"><span>ego<span style="color:#f92672">&lt;-</span><span style="color:#a6e22e">read.table</span>(<span style="color:#e6db74">&#34;GO.txt&#34;</span>,sep<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;\t&#34;</span>,check.names<span style="color:#f92672">=</span>F,header<span style="color:#f92672">=</span>T)
</span></span><span style="display:flex;"><span>go<span style="color:#f92672">=</span><span style="color:#a6e22e">data.frame</span>(Category <span style="color:#f92672">=</span><span style="color:#e6db74">&#34;ALL&#34;</span>,ID <span style="color:#f92672">=</span> ego<span style="color:#f92672">$</span>ID,Term <span style="color:#f92672">=</span> ego<span style="color:#f92672">$</span>Description, Genes <span style="color:#f92672">=</span> <span style="color:#a6e22e">gsub</span>(<span style="color:#e6db74">&#34;/&#34;</span>, <span style="color:#e6db74">&#34;, &#34;</span>, ego<span style="color:#f92672">$</span>geneID), adj_pval <span style="color:#f92672">=</span> ego<span style="color:#f92672">$</span>pvalue)
</span></span><span style="display:flex;"><span>id.fc<span style="color:#f92672">=</span>rt
</span></span><span style="display:flex;"><span>genelist <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">data.frame</span>(ID <span style="color:#f92672">=</span> id.fc<span style="color:#f92672">$</span>gene, logFC <span style="color:#f92672">=</span> id.fc<span style="color:#f92672">$</span>logFC)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">row.names</span>(genelist)<span style="color:#f92672">=</span>genelist[,<span style="color:#ae81ff">1</span>]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>circ <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">circle_dat</span>(go, genelist)
</span></span><span style="display:flex;"><span>termNum <span style="color:#f92672">=</span> <span style="color:#ae81ff">5</span>                                     <span style="color:#75715e">#限定term数目</span>
</span></span><span style="display:flex;"><span>geneNum <span style="color:#f92672">=</span> <span style="color:#a6e22e">nrow</span>(genelist)                        <span style="color:#75715e">#限定基因数目可以改为数字</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>chord <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">chord_dat</span>(circ, genelist[1<span style="color:#f92672">:</span>geneNum,], go<span style="color:#f92672">$</span>Term[1<span style="color:#f92672">:</span>termNum])
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">pdf</span>(file<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;GO_circ.pdf&#34;</span>,width <span style="color:#f92672">=</span> <span style="color:#ae81ff">12</span>,height <span style="color:#f92672">=</span> <span style="color:#ae81ff">11</span>)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">GOChord</span>(chord, 
</span></span><span style="display:flex;"><span>        space <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.001</span>,           <span style="color:#75715e">#基因之间的间距</span>
</span></span><span style="display:flex;"><span>        gene.order <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;logFC&#39;</span>,    <span style="color:#75715e">#按照logFC值对基因排序</span>
</span></span><span style="display:flex;"><span>        gene.space <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.25</span>,       <span style="color:#75715e">#基因名跟圆圈的相对距离</span>
</span></span><span style="display:flex;"><span>        gene.size <span style="color:#f92672">=</span> <span style="color:#ae81ff">5</span>,           <span style="color:#75715e">#基因名字体大小 </span>
</span></span><span style="display:flex;"><span>        border.size <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.1</span>,       <span style="color:#75715e">#线条粗细</span>
</span></span><span style="display:flex;"><span>        process.label <span style="color:#f92672">=</span> <span style="color:#ae81ff">9</span>)       <span style="color:#75715e">#term字体大小</span>
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">dev.off</span>()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#75715e">####KEGG####</span>
</span></span><span style="display:flex;"><span>rt<span style="color:#f92672">=</span><span style="color:#a6e22e">read.table</span>(<span style="color:#e6db74">&#34;input.txt&#34;</span>,sep<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;\t&#34;</span>,check.names<span style="color:#f92672">=</span>F,header<span style="color:#f92672">=</span>T)    <span style="color:#75715e">#读取文件</span>
</span></span><span style="display:flex;"><span>genes<span style="color:#f92672">=</span><span style="color:#a6e22e">as.vector</span>(rt[,<span style="color:#ae81ff">1</span>])
</span></span><span style="display:flex;"><span>entrezIDs <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">mget</span>(genes, org.Hs.egSYMBOL2EG, ifnotfound<span style="color:#f92672">=</span><span style="color:#66d9ef">NA</span>)    <span style="color:#75715e">#找出基因对应的id</span>
</span></span><span style="display:flex;"><span>entrezIDs <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">as.character</span>(entrezIDs)
</span></span><span style="display:flex;"><span>out<span style="color:#f92672">=</span><span style="color:#a6e22e">cbind</span>(rt,entrezID<span style="color:#f92672">=</span>entrezIDs)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">write.table</span>(out,file<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;KEGG-id.txt&#34;</span>,sep<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;\t&#34;</span>,quote<span style="color:#f92672">=</span>F,row.names<span style="color:#f92672">=</span>F)    <span style="color:#75715e">#输出结果</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>rt<span style="color:#f92672">=</span><span style="color:#a6e22e">read.table</span>(<span style="color:#e6db74">&#34;KEGG-id.txt&#34;</span>,sep<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;\t&#34;</span>,header<span style="color:#f92672">=</span>T,check.names<span style="color:#f92672">=</span>F)       
</span></span><span style="display:flex;"><span>rt<span style="color:#f92672">=</span>rt<span style="color:#a6e22e">[is.na</span>(rt[,<span style="color:#e6db74">&#34;entrezID&#34;</span>])<span style="color:#f92672">==</span>F,]                             
</span></span><span style="display:flex;"><span>gene<span style="color:#f92672">=</span>rt<span style="color:#f92672">$</span>entrezID
</span></span><span style="display:flex;"><span><span style="color:#75715e">#kegg分析</span>
</span></span><span style="display:flex;"><span>kk <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">enrichKEGG</span>(gene <span style="color:#f92672">=</span> gene,keyType <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;kegg&#34;</span>,organism <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;hsa&#34;</span>, pvalueCutoff <span style="color:#f92672">=</span><span style="color:#ae81ff">0.05</span>, qvalueCutoff <span style="color:#f92672">=</span><span style="color:#ae81ff">0.05</span>, pAdjustMethod <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;fdr&#34;</span>)   
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">write.table</span>(kk,file<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;KEGG.txt&#34;</span>,sep<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;\t&#34;</span>,quote<span style="color:#f92672">=</span>F,row.names <span style="color:#f92672">=</span> F)                         
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#75715e">#可视化</span>
</span></span><span style="display:flex;"><span><span style="color:#75715e">##条形图</span>
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">pdf</span>(file<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;KEGG-barplot.pdf&#34;</span>,width <span style="color:#f92672">=</span> <span style="color:#ae81ff">10</span>,height <span style="color:#f92672">=</span> <span style="color:#ae81ff">13</span>)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">barplot</span>(kk, drop <span style="color:#f92672">=</span> <span style="color:#66d9ef">TRUE</span>, showCategory <span style="color:#f92672">=</span> <span style="color:#ae81ff">15</span>,label_format<span style="color:#f92672">=</span><span style="color:#ae81ff">100</span>)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">dev.off</span>()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#75715e">##气泡图</span>
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">pdf</span>(file<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;KEGG-bubble.pdf&#34;</span>,width <span style="color:#f92672">=</span> <span style="color:#ae81ff">10</span>,height <span style="color:#f92672">=</span> <span style="color:#ae81ff">13</span>)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">dotplot</span>(kk, showCategory <span style="color:#f92672">=</span> <span style="color:#ae81ff">15</span>,label_format<span style="color:#f92672">=</span><span style="color:#ae81ff">100</span>)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">dev.off</span>()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#75715e">#圈图</span>
</span></span><span style="display:flex;"><span>ego<span style="color:#f92672">&lt;-</span><span style="color:#a6e22e">read.table</span>(<span style="color:#e6db74">&#34;KEGG.txt&#34;</span>,sep<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;\t&#34;</span>,check.names<span style="color:#f92672">=</span>F,header<span style="color:#f92672">=</span>T)
</span></span><span style="display:flex;"><span>go<span style="color:#f92672">=</span><span style="color:#a6e22e">data.frame</span>(Category <span style="color:#f92672">=</span><span style="color:#e6db74">&#34;ALL&#34;</span>,ID <span style="color:#f92672">=</span> ego<span style="color:#f92672">$</span>ID,Term <span style="color:#f92672">=</span> ego<span style="color:#f92672">$</span>Description, Genes <span style="color:#f92672">=</span> <span style="color:#a6e22e">gsub</span>(<span style="color:#e6db74">&#34;/&#34;</span>, <span style="color:#e6db74">&#34;, &#34;</span>, ego<span style="color:#f92672">$</span>geneID), adj_pval <span style="color:#f92672">=</span> ego<span style="color:#f92672">$</span>p.adjust)
</span></span><span style="display:flex;"><span>id.fc<span style="color:#f92672">=</span>rt
</span></span><span style="display:flex;"><span>genelist <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">data.frame</span>(ID <span style="color:#f92672">=</span> id.fc<span style="color:#f92672">$</span>entrezID, logFC <span style="color:#f92672">=</span> id.fc<span style="color:#f92672">$</span>logFC)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">row.names</span>(genelist)<span style="color:#f92672">=</span>genelist[,<span style="color:#ae81ff">1</span>]
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">row.names</span>(rt)<span style="color:#f92672">=</span>rt[,<span style="color:#ae81ff">3</span>]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>circ <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">circle_dat</span>(go, genelist)
</span></span><span style="display:flex;"><span>termNum <span style="color:#f92672">=</span> <span style="color:#ae81ff">5</span>                                     <span style="color:#75715e">#限定term数目</span>
</span></span><span style="display:flex;"><span>geneNum <span style="color:#f92672">=</span> <span style="color:#a6e22e">nrow</span>(genelist)                        <span style="color:#75715e">#限定基因数目可以改为数字</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>chord <span style="color:#f92672">&lt;-</span> <span style="color:#a6e22e">chord_dat</span>(circ, genelist[1<span style="color:#f92672">:</span>geneNum,], go<span style="color:#f92672">$</span>Term[1<span style="color:#f92672">:</span>termNum])
</span></span><span style="display:flex;"><span>sameSample<span style="color:#f92672">=</span><span style="color:#a6e22e">intersect</span>(<span style="color:#a6e22e">row.names</span>(chord), <span style="color:#a6e22e">row.names</span>(rt))
</span></span><span style="display:flex;"><span>rt<span style="color:#f92672">=</span>rt[sameSample,,drop<span style="color:#f92672">=</span>F]
</span></span><span style="display:flex;"><span>geneIDs<span style="color:#f92672">=</span>rt<span style="color:#f92672">$</span>gene
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">row.names</span>(chord)<span style="color:#f92672">=</span>geneIDs
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">pdf</span>(file<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;KEGG_circ.pdf&#34;</span>,width <span style="color:#f92672">=</span> <span style="color:#ae81ff">12</span>,height <span style="color:#f92672">=</span> <span style="color:#ae81ff">11</span>)
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">GOChord</span>(chord, 
</span></span><span style="display:flex;"><span>        space <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.001</span>,           <span style="color:#75715e">#基因之间的间距</span>
</span></span><span style="display:flex;"><span>        gene.order <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;logFC&#39;</span>,    <span style="color:#75715e">#按照logFC值对基因排序</span>
</span></span><span style="display:flex;"><span>        gene.space <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.25</span>,       <span style="color:#75715e">#基因名跟圆圈的相对距离</span>
</span></span><span style="display:flex;"><span>        gene.size <span style="color:#f92672">=</span> <span style="color:#ae81ff">5</span>,           <span style="color:#75715e">#基因名字体大小 </span>
</span></span><span style="display:flex;"><span>        border.size <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.1</span>,       <span style="color:#75715e">#线条粗细</span>
</span></span><span style="display:flex;"><span>        process.label <span style="color:#f92672">=</span> <span style="color:#ae81ff">9</span>)       <span style="color:#75715e">#term字体大小</span>
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">dev.off</span>()
</span></span></code></pre></div>
        
    </div>

    

</article>
          
            <article class="post-preview">
    <a href="https://cywd123.github.io/posts/%E5%AD%9F%E5%BE%B7%E5%B0%94%E9%9A%8F%E6%9C%BA%E5%8C%96%E5%AD%A6%E4%B9%A0%E7%AC%94%E8%AE%B0%E5%9B%9B/">
        <h2 class="post-title">孟德尔随机化笔记（四）——药物靶点MR分析</h2>
        
        
        
    </a>

    <p class="post-meta">
        <span class="post-meta">
  
  
  <i class="fas fa-calendar"></i>&nbsp;Posted on August 5, 2023
  
    &nbsp;(Last modified on December 1, 2025)
  
  
  
  
    
      &nbsp;|&nbsp;<i class="fas fa-user"></i>&nbsp;
    
  
  
</span>


    </p>
    <div class="post-entry">
        
        <p>用于药物开发与效应预测的药物孟德尔随机化（Drug-MR）,基于靶蛋白的下游产物（biomarker），以靶蛋白编码基因附近的对biomarker有显著效应的SNP（pQTL或者eQTL）作为工具变量，以biomarker浓度作为暴露，以疾病作为结局，进行孟德尔随机化，以验证蛋白靶对于所研究疾病的影响。</p>
<p><a href="https://zhuanlan.zhihu.com/p/548337552">参考资料</a></p>
<p>资料建议剔除连锁不平衡： R^2 = 0.60</p>
<p>下面是分别用easyMR和MendelR包进行的分析代码</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#a6e22e">library</span>(easyMR)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>dat<span style="color:#f92672">=</span><span style="color:#a6e22e">get_drug_target_data</span>(
</span></span><span style="display:flex;"><span>  id <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;ieu-b-110&#34;</span>,　　　　<span style="color:#75715e">#暴露GWAS ID</span>
</span></span><span style="display:flex;"><span>  gene_name<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;HMGCR&#34;</span>,      <span style="color:#75715e">#药靶蛋白编码基因</span>
</span></span><span style="display:flex;"><span>  kb<span style="color:#f92672">=</span><span style="color:#ae81ff">100</span>,          <span style="color:#75715e">#基因附近的SNP范围</span>
</span></span><span style="display:flex;"><span>  clump_kb <span style="color:#f92672">=</span> <span style="color:#ae81ff">100</span>,   <span style="color:#75715e">#clump的范围</span>
</span></span><span style="display:flex;"><span>  clump_local <span style="color:#f92672">=</span> <span style="color:#66d9ef">FALSE</span>,
</span></span><span style="display:flex;"><span>  r2<span style="color:#f92672">=</span><span style="color:#ae81ff">0.3</span>,   <span style="color:#75715e">#clump的r2阈值</span>
</span></span><span style="display:flex;"><span>  pval<span style="color:#f92672">=</span><span style="color:#ae81ff">0.05</span>,
</span></span><span style="display:flex;"><span>  MAF <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.01</span>,  <span style="color:#75715e">#次等位基因频率 阈值</span>
</span></span><span style="display:flex;"><span>  build <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;GRch38&#34;</span>,
</span></span><span style="display:flex;"><span>  chr <span style="color:#f92672">=</span> <span style="color:#66d9ef">NULL</span>,
</span></span><span style="display:flex;"><span>  pos_start <span style="color:#f92672">=</span> <span style="color:#66d9ef">NULL</span>,
</span></span><span style="display:flex;"><span>  pos_end <span style="color:#f92672">=</span> <span style="color:#66d9ef">NULL</span>
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">drug_MR</span>(
</span></span><span style="display:flex;"><span>  target_gene_data<span style="color:#f92672">=</span>dat,
</span></span><span style="display:flex;"><span>  outcome_id <span style="color:#f92672">=</span><span style="color:#e6db74">&#34;finn-b-I9_AF&#34;</span>,  <span style="color:#75715e">#疾病GWAS ID</span>
</span></span><span style="display:flex;"><span>  outcome_name<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;atrial fibrillation&#34;</span>,
</span></span><span style="display:flex;"><span>  inhibitor <span style="color:#f92672">=</span> <span style="color:#66d9ef">FALSE</span>,
</span></span><span style="display:flex;"><span>  after_trans_inhibitor <span style="color:#f92672">=</span> <span style="color:#66d9ef">FALSE</span>,
</span></span><span style="display:flex;"><span>  pval <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.05</span>,
</span></span><span style="display:flex;"><span>  action <span style="color:#f92672">=</span> <span style="color:#ae81ff">2</span>,
</span></span><span style="display:flex;"><span>  out_type <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;binary&#34;</span>,
</span></span><span style="display:flex;"><span>  save_path<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;e:&#34;</span>
</span></span><span style="display:flex;"><span>)
</span></span></code></pre></div><div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#a6e22e">library</span>(MendelR)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">mr_common</span>(
</span></span><span style="display:flex;"><span>  id_exposure <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;ieu-b-110&#34;</span>,  <span style="color:#75715e">#暴露GWAS ID</span>
</span></span><span style="display:flex;"><span>  id_outcome <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;finn-b-I9_AF&#34;</span>,  <span style="color:#75715e">#疾病GWAS ID</span>
</span></span><span style="display:flex;"><span>  p1 <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.05</span>,
</span></span><span style="display:flex;"><span>  p2 <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.05</span>,
</span></span><span style="display:flex;"><span>  write_csv <span style="color:#f92672">=</span> <span style="color:#66d9ef">TRUE</span>,
</span></span><span style="display:flex;"><span>  write_ppt <span style="color:#f92672">=</span> <span style="color:#66d9ef">FALSE</span>,
</span></span><span style="display:flex;"><span>  method_list <span style="color:#f92672">=</span> <span style="color:#a6e22e">c</span>(<span style="color:#e6db74">&#34;mr_ivw&#34;</span>, <span style="color:#e6db74">&#34;mr_egger_regression&#34;</span>, <span style="color:#e6db74">&#34;mr_weighted_median&#34;</span>,
</span></span><span style="display:flex;"><span>    <span style="color:#e6db74">&#34;mr_weighted_mode&#34;</span>),
</span></span><span style="display:flex;"><span>  rm_snps <span style="color:#f92672">=</span> <span style="color:#66d9ef">NULL</span>,
</span></span><span style="display:flex;"><span>  r2 <span style="color:#f92672">=</span> <span style="color:#ae81ff">0.3</span>,  <span style="color:#75715e">#clump的r2阈值</span>
</span></span><span style="display:flex;"><span>  kb <span style="color:#f92672">=</span> <span style="color:#ae81ff">100</span>,  <span style="color:#75715e">#clump的范围</span>
</span></span><span style="display:flex;"><span>  build_version <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;hg19&#34;</span>,
</span></span><span style="display:flex;"><span>  gene <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;HMGCR&#34;</span>,     <span style="color:#75715e">#药靶蛋白编码基因</span>
</span></span><span style="display:flex;"><span>  chr <span style="color:#f92672">=</span> <span style="color:#66d9ef">NULL</span>,
</span></span><span style="display:flex;"><span>  pos_start <span style="color:#f92672">=</span> <span style="color:#66d9ef">NULL</span>,
</span></span><span style="display:flex;"><span>  pos_end <span style="color:#f92672">=</span> <span style="color:#66d9ef">NULL</span>,
</span></span><span style="display:flex;"><span>  eaf_threshold <span style="color:#f92672">=</span> <span style="color:#66d9ef">NULL</span>,
</span></span><span style="display:flex;"><span>  run_presso <span style="color:#f92672">=</span> T,
</span></span><span style="display:flex;"><span>  gene_win <span style="color:#f92672">=</span> <span style="color:#ae81ff">100</span>,   <span style="color:#75715e">#基因附近的SNP范围</span>
</span></span><span style="display:flex;"><span>  NbDistribution <span style="color:#f92672">=</span> <span style="color:#ae81ff">3000</span>,
</span></span><span style="display:flex;"><span>  find_proxy <span style="color:#f92672">=</span> T,
</span></span><span style="display:flex;"><span>  local_clump <span style="color:#f92672">=</span> F,
</span></span><span style="display:flex;"><span>  r2_cal_mode <span style="color:#f92672">=</span> <span style="color:#ae81ff">1</span>,
</span></span><span style="display:flex;"><span>  steiger <span style="color:#f92672">=</span> T,   <span style="color:#75715e">#是否进行Steiger方向性检验</span>
</span></span><span style="display:flex;"><span>  auto_ivw <span style="color:#f92672">=</span> T,
</span></span><span style="display:flex;"><span>  pop <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;EUR&#34;</span>,
</span></span><span style="display:flex;"><span>  no_clump <span style="color:#f92672">=</span> F,
</span></span><span style="display:flex;"><span>  out_dir <span style="color:#f92672">=</span> <span style="color:#66d9ef">NULL</span>,
</span></span><span style="display:flex;"><span>  exposure_samplesize <span style="color:#f92672">=</span> <span style="color:#66d9ef">NULL</span>,
</span></span><span style="display:flex;"><span>  outcome_samplesize <span style="color:#f92672">=</span> <span style="color:#66d9ef">NULL</span>
</span></span><span style="display:flex;"><span>)
</span></span></code></pre></div>
        
    </div>

    

</article>
          
            <article class="post-preview">
    <a href="https://cywd123.github.io/posts/%E6%9C%BA%E5%99%A8%E5%AD%A6%E4%B9%A012%E7%A7%8D%E5%88%86%E7%B1%BB%E6%A8%A1%E5%9E%8B%E7%9A%84%E6%80%A7%E8%83%BD%E5%AF%B9%E6%AF%94/">
        <h2 class="post-title">机器学习12种分类模型的性能对比</h2>
        
        
        
    </a>

    <p class="post-meta">
        <span class="post-meta">
  
  
  <i class="fas fa-calendar"></i>&nbsp;Posted on June 19, 2023
  
    &nbsp;(Last modified on December 1, 2025)
  
  
  
  
    
      &nbsp;|&nbsp;<i class="fas fa-user"></i>&nbsp;
    
  
  
</span>


    </p>
    <div class="post-entry">
        
        <p>接上回，我们已经知道了<a href="/2023/05/21/%E6%9C%BA%E5%99%A8%E5%AD%A6%E4%B9%A0%E5%88%86%E7%B1%BB%E7%9A%84%E6%9C%AC%E8%B4%A8/">机器学习分类的本质</a>，那么接下来就是要介绍机器学习中的分类模型的各种性能了。</p>
<h2 id="简介">简介</h2>
<p>我们将采用12种分类模型，分别是：GaussianNB、MultinomialNB、KNNeighbors、SVC、DecisionTree、RandomForest、GradientBoosting、LGBM、XGB、CatBoost、AdaBoost、MLP，用这些模型对数据集进行训练，然后对测试集进行预测，最后将预测结果与真实结果进行对比，得到各个模型的性能。</p>
<h2 id="数据集">数据集</h2>
<p>我们采用以下代码生成随机分类数据：</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>x,y <span style="color:#f92672">=</span> make_classification(n_samples<span style="color:#f92672">=</span><span style="color:#ae81ff">100000</span>,n_features<span style="color:#f92672">=</span><span style="color:#ae81ff">3</span>,n_classes<span style="color:#f92672">=</span><span style="color:#ae81ff">4</span>,n_informative<span style="color:#f92672">=</span><span style="color:#ae81ff">3</span>,n_redundant<span style="color:#f92672">=</span><span style="color:#ae81ff">0</span>,random_state<span style="color:#f92672">=</span><span style="color:#ae81ff">50</span>,n_clusters_per_class<span style="color:#f92672">=</span><span style="color:#ae81ff">1</span>)
</span></span><span style="display:flex;"><span>x_train,x_test,y_train,y_test <span style="color:#f92672">=</span> train_test_split(x,y,test_size<span style="color:#f92672">=</span><span style="color:#ae81ff">0.1</span>)
</span></span></code></pre></div><p>这里能够生成100000个样本，每个样本有3个特征，4个类别，3个特征是有效特征，0个冗余特征，随机种子为50，每个类别有1个簇。然后我们将数据集分为训练集和测试集，测试集占10%.<br>
由于每个样本都是一个3维向量，所以我们将数据取前2维数据将数据集可视化，如下图所示：<br>
<img src="/images/1741977356661.png" alt=""></p>
<h2 id="评估函数">评估函数</h2>
<p>我们采用准确率(Accuracy)、精确率(Precision)、召回率(Recall)、F1值(F1-score)、R2值(R2-score)，这5个指标来评估模型的性能。</p>
<h4 id="准确率accuracy">准确率(Accuracy)</h4>
<p>准确率是指分类正确的样本数占总样本数的比例，即：</p>
<p>$Accuracy = \frac{TP+TN}{TP+TN+FP+FN}$</p>
<p>其中，TP是真正例，TN是真负例，FP是假正例，FN是假负例。</p>
<h4 id="精确率precision">精确率(Precision)</h4>
<p>精确率是指分类正确的正例数占分类为正例的样本数的比例，即：</p>
<!-- raw HTML omitted -->
<p>其中，TP是真正例，FP是假正例。</p>
<h4 id="召回率recall">召回率(Recall)</h4>
<p>召回率是指分类正确的正例数占真正例的比例，即：</p>
<p>$Recall = \frac{TP}{TP+FN}$</p>
<p>其中，TP是真正例，FN是假负例。</p>
<h4 id="f1值f1-score">F1值(F1-score)</h4>
<p>F1值是精确率和召回率的调和平均数，即：</p>
<!-- raw HTML omitted -->
<p>F1值越接近1，表明模型的性能越好。</p>
<h4 id="r2值r2-score">R2值(R2-score)</h4>
<p>R2值是指预测值与真实值的相关系数，即：</p>
<!-- raw HTML omitted -->
<p>其中，$y_i$是真实值，$\hat{y_i}$是预测值，$\bar{y}$是真实值的均值。</p>
<p>R2值越接近1，表明模型的性能越好。</p>
<h2 id="各模型表现">各模型表现</h2>
<h3 id="1-gaussiannb">1. GaussianNB</h3>
<p><img src="/images/1741977363429.png" alt=""><br>
其中GaussianNB的参数priors为None，var_smoothing为1e-09，表明我们没有对先验概率进行设置，而且我们对方差进行了平滑处理。</p>
<h3 id="2-multinomialnb">2. MultinomialNB</h3>
<p><img src="/images/1741977369592.png" alt=""><br>
其中MultinomialNB的参数alpha为1.0，fit_prior为True，class_prior为None，表明我们对先验概率进行了设置，而且我们对先验概率进行了平滑处理。</p>
<h3 id="3-knneighbors">3. KNNeighbors</h3>
<p><img src="/images/1741977374241.png" alt=""><br>
其中KNNeighbors的参数n_neighbors为30，表明我们对最近邻的个数为30。</p>
<h3 id="4-svc">4. SVC</h3>
<p><img src="/images/1741977380654.png" alt=""></p>
<h3 id="5-decisiontree">5. DecisionTree</h3>
<p><img src="/images/1741977386093.png" alt=""></p>
<h3 id="6-randomforest">6. RandomForest</h3>
<p><img src="/images/1741977391536.png" alt=""></p>
        <a href="https://cywd123.github.io/posts/%E6%9C%BA%E5%99%A8%E5%AD%A6%E4%B9%A012%E7%A7%8D%E5%88%86%E7%B1%BB%E6%A8%A1%E5%9E%8B%E7%9A%84%E6%80%A7%E8%83%BD%E5%AF%B9%E6%AF%94/" class="post-read-more">[Read More]</a>
        
    </div>

    

</article>
          
            <article class="post-preview">
    <a href="https://cywd123.github.io/posts/%E6%9C%BA%E5%99%A8%E5%AD%A6%E4%B9%A0%E5%88%86%E7%B1%BB%E7%9A%84%E6%9C%AC%E8%B4%A8/">
        <h2 class="post-title">机器学习分类的本质</h2>
        
        
        
    </a>

    <p class="post-meta">
        <span class="post-meta">
  
  
  <i class="fas fa-calendar"></i>&nbsp;Posted on May 21, 2023
  
    &nbsp;(Last modified on December 1, 2025)
  
  
  
  
    
      &nbsp;|&nbsp;<i class="fas fa-user"></i>&nbsp;
    
  
  
</span>


    </p>
    <div class="post-entry">
        
        <h1 id="机器学习分类的本质">机器学习分类的本质</h1>
<h2 id="引入">引入</h2>
<p>机器学习中，基本的任务可以分为三类：分类、回归和聚类。</p>
<p>分类和回归的区别在于，分类的输出是离散的，而回归的输出是连续的。聚类的任务是将数据集中的样本分成若干个类别，每个类别称为一个簇，簇内的样本相似度较高，而簇间的样本相似度较低。其中，分类和回归的任务是有监督学习，而聚类的任务是无监督学习。</p>
<h2 id="分类问题">分类问题</h2>
<p>本次主要介绍分类问题，它属于监督学习的范畴。</p>
<p>分类的本质是学习一个分类函数，将输入空间映射到输出空间，即：$f: \mathcal{X} \rightarrow \mathcal{Y}$，其中，$\mathcal{X}$是输入空间，$\mathcal{Y}$是输出空间。</p>
<p>一般来说，输入空间是由特征向量构成的，即：$\mathcal{X} = \mathbb{R}^n$，输出空间是离散的，即：$\mathcal{Y} = {c_1, c_2, \cdots, c_k}$，其中，$c_i$是类别标签。</p>
<p>特征向量是由特征构成的，就是对样本的描述，是样本的某个属性。</p>
<p>这里面就可以构建给出两个特征向量：</p>
<p>（5.1，3.5，1.4，0.2）、（4.9，3，1.4，0.2）</p>
<hr>
<blockquote>
<p>特征的选择对分类的性能有很大的影响，特征的选择应该具有以下几个特点：</p>
<ol>
<li>
<p>特征应该能够很好地区分不同类别的样本。</p>
</li>
<li>
<p>特征应该具有可解释性，即：特征应该能够很好地解释样本的类别。</p>
</li>
<li>
<p>特征应该具有鲁棒性。（泛化能力）</p>
</li>
<li>
<p>特征应该具有可扩展性，能够很好地扩展到新的样本。</p>
</li>
<li>
<p>特征应该具有可计算性。</p>
</li>
<li>
<p>特征应该具有低维性。</p>
</li>
</ol>
</blockquote>
<hr>
<p>输出空间是离散的，即：$\mathcal{Y} = {c_1, c_2, \cdots, c_k}$，其中，$c_i$是类别标签。</p>
<p>在我接触的的深度学习分类中，一般将输出空间定义为独热编码（One-Hot Encoding）的形式，即：$\mathcal{Y} = {[1, 0, \cdots, 0], [0, 1, \cdots, 0], \cdots, [0, 0, \cdots, 1]}$，其中，$[1, 0, \cdots, 0]$表示类别$c_1$，$[0, 1, \cdots, 0]$表示类别$c_2$，以此类推。</p>
<hr>
<blockquote>
<p>独热编码是一种常用的编码方式，它将离散的类别标签转换为离散的向量，其中，向量的维度等于类别的个数，向量的值等于类别的索引。独热编码的好处是，它能够很好地表示类别之间的关系，而且它的值是离散的，不会产生类别之间的大小关系。</p>
</blockquote>
<hr>
<p>通常机器学习中输出的结果一般不是只有0和1，而是在0~1之间的小数，这个小数表示样本属于某个类别的概率，即：$P(Y=c_i|X)$，其中，$c_i$是类别标签，$X$是特征向量。</p>
<h2 id="分类的方法">分类的方法</h2>
<p>分类的方法主要分为两类：生成方法和判别方法。</p>
<h3 id="生成方法">生成方法</h3>
<p>生成方法是通过学习联合概率分布$P(X, Y)$来进行分类的，即：$P(Y|X) = \frac{P(X, Y)}{P(X)}$，其中，$P(Y|X)$是后验概率，$P(X)$是先验概率，$P(X, Y)$是联合概率分布。</p>
<h4 id="朴素贝叶斯">朴素贝叶斯</h4>
<p>朴素贝叶斯是一种生成方法，它假设特征之间相互独立，即：$P(X|Y) = \prod_{i=1}^n P(x_i|Y)$，其中，$x_i$是特征向量的第$i$个特征。</p>
        <a href="https://cywd123.github.io/posts/%E6%9C%BA%E5%99%A8%E5%AD%A6%E4%B9%A0%E5%88%86%E7%B1%BB%E7%9A%84%E6%9C%AC%E8%B4%A8/" class="post-read-more">[Read More]</a>
        
    </div>

    

</article>
          
            <article class="post-preview">
    <a href="https://cywd123.github.io/posts/%E8%AE%BA%E6%96%87%E4%BD%A0%E5%A5%BD%E8%AE%BA%E6%96%87%E5%86%8D%E8%A7%81/">
        <h2 class="post-title">论文📃你好，论文📃再见。</h2>
        
        
        
    </a>

    <p class="post-meta">
        <span class="post-meta">
  
  
  <i class="fas fa-calendar"></i>&nbsp;Posted on April 15, 2023
  
    &nbsp;(Last modified on December 1, 2025)
  
  
  
  
    
      &nbsp;|&nbsp;<i class="fas fa-user"></i>&nbsp;
    
  
  
</span>


    </p>
    <div class="post-entry">
        
        <p>论文📃你好，论文📃再见。<br>
<img src="/images/1741977163120.jpg" alt=""></p>

        
    </div>

    

</article>
          
            <article class="post-preview">
    <a href="https://cywd123.github.io/posts/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0%E6%89%8B%E5%86%99%E6%95%B0%E5%AD%97%E5%9B%BE%E5%83%8F%E8%AF%86%E5%88%AB%E5%88%86%E7%B1%BBmnist%E6%95%B0%E6%8D%AE%E9%9B%86/">
        <h2 class="post-title">【深度学习】手写数字图像识别分类（MNIST数据集）</h2>
        
        
        
    </a>

    <p class="post-meta">
        <span class="post-meta">
  
  
  <i class="fas fa-calendar"></i>&nbsp;Posted on April 1, 2023
  
    &nbsp;(Last modified on December 1, 2025)
  
  
  
  
    
      &nbsp;|&nbsp;<i class="fas fa-user"></i>&nbsp;
    
  
  
</span>


    </p>
    <div class="post-entry">
        
        <h1 id="卷积神经网络">卷积神经网络</h1>
<h2 id="mnist介绍">MNIST介绍</h2>
<p>MNIST 数据集可在 <a href="http://yann.lecun.com/exdb/mnist/">http://yann.lecun.com/exdb/mnist/</a> 获取, 它包含了四个部分:</p>
<p>Training set images: train-images-idx3-ubyte.gz (9.9 MB, 解压后 47 MB, 包含 60,000 个样本)<br>
Training set labels: train-labels-idx1-ubyte.gz (29 KB, 解压后 60 KB, 包含 60,000 个标签)<br>
Test set images: t10k-images-idx3-ubyte.gz (1.6 MB, 解压后 7.8 MB, 包含 10,000 个样本)<br>
Test set labels: t10k-labels-idx1-ubyte.gz (5KB, 解压后 10 KB, 包含 10,000 个标签)</p>
<p>MNIST 数据集来自美国国家标准与技术研究所, National Institute of Standards and Technology (NIST). 训练集 (training set) 由来自 250 个不同人手写的数字构成, 其中 50% 是高中学生, 50% 来自人口普查局 (the Census Bureau) 的工作人员. 测试集(test set) 也是同样比例的手写数字数据.</p>
        <a href="https://cywd123.github.io/posts/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0%E6%89%8B%E5%86%99%E6%95%B0%E5%AD%97%E5%9B%BE%E5%83%8F%E8%AF%86%E5%88%AB%E5%88%86%E7%B1%BBmnist%E6%95%B0%E6%8D%AE%E9%9B%86/" class="post-read-more">[Read More]</a>
        
    </div>

    
    <div class="blog-tags">
        
       
       <a href="https://cywd123.github.io/tags/%E7%BC%96%E7%A8%8B/">编程</a>&nbsp;
        
       
       <a href="https://cywd123.github.io/tags/%E6%B7%B1%E5%BA%A6%E5%AD%A6%E4%B9%A0/">深度学习</a>&nbsp;
         
    </div>
    

</article>
          
            <article class="post-preview">
    <a href="https://cywd123.github.io/posts/%E6%9C%BA%E5%99%A8%E5%AD%A6%E4%B9%A0%E5%9B%9B%E5%B1%82%E4%BA%BA%E5%B7%A5%E7%A5%9E%E7%BB%8F%E7%BD%91%E7%BB%9C%E6%8B%9F%E5%90%88%E4%BA%8C%E5%85%83%E4%BA%8C%E6%AC%A1%E5%87%BD%E6%95%B0/">
        <h2 class="post-title">【机器学习】四层人工神经网络拟合二元二次函数</h2>
        
        
        
    </a>

    <p class="post-meta">
        <span class="post-meta">
  
  
  <i class="fas fa-calendar"></i>&nbsp;Posted on March 29, 2023
  
    &nbsp;(Last modified on December 1, 2025)
  
  
  
  
    
      &nbsp;|&nbsp;<i class="fas fa-user"></i>&nbsp;
    
  
  
</span>


    </p>
    <div class="post-entry">
        
        <h1 id="四层人工神经网络拟合二元二次函数">四层人工神经网络拟合二元二次函数</h1>
<p>如果想了解人工神经网络，请认真阅读每个字，字字珠玑。😂</p>
<p>核心代码:</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>      <span style="color:#960050;background-color:#1e0010">、、、、、、、、、、</span>
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">class</span> <span style="color:#a6e22e">Net</span>(nn<span style="color:#f92672">.</span>Module):
</span></span><span style="display:flex;"><span>    <span style="color:#66d9ef">def</span> <span style="color:#a6e22e">__init__</span>(self):
</span></span><span style="display:flex;"><span>        super(Net, self)<span style="color:#f92672">.</span><span style="color:#a6e22e">__init__</span>()
</span></span><span style="display:flex;"><span>        self<span style="color:#f92672">.</span>fc1 <span style="color:#f92672">=</span> nn<span style="color:#f92672">.</span>Linear(<span style="color:#ae81ff">2</span>, <span style="color:#ae81ff">40</span>)
</span></span><span style="display:flex;"><span>        self<span style="color:#f92672">.</span>fc2 <span style="color:#f92672">=</span> nn<span style="color:#f92672">.</span>Linear(<span style="color:#ae81ff">40</span>, <span style="color:#ae81ff">40</span>)
</span></span><span style="display:flex;"><span>        self<span style="color:#f92672">.</span>fc3 <span style="color:#f92672">=</span>nn<span style="color:#f92672">.</span>Linear(<span style="color:#ae81ff">40</span>, <span style="color:#ae81ff">24</span>)
</span></span><span style="display:flex;"><span>        self<span style="color:#f92672">.</span>fc4 <span style="color:#f92672">=</span>nn<span style="color:#f92672">.</span>Linear(<span style="color:#ae81ff">24</span>, <span style="color:#ae81ff">1</span>)
</span></span><span style="display:flex;"><span>    <span style="color:#66d9ef">def</span> <span style="color:#a6e22e">forward</span>(self,x):
</span></span><span style="display:flex;"><span>        <span style="color:#66d9ef">return</span> self<span style="color:#f92672">.</span>fc4(t<span style="color:#f92672">.</span>relu(self<span style="color:#f92672">.</span>fc3(t<span style="color:#f92672">.</span>relu(self<span style="color:#f92672">.</span>fc2(t<span style="color:#f92672">.</span>relu(self<span style="color:#f92672">.</span>fc1(x)))))))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">def</span> <span style="color:#a6e22e">train</span>(epoch,model,opt,lossfn,tu,tc):
</span></span><span style="display:flex;"><span>    <span style="color:#66d9ef">for</span> i <span style="color:#f92672">in</span> range(<span style="color:#ae81ff">1</span>,epoch<span style="color:#f92672">+</span><span style="color:#ae81ff">1</span>):
</span></span><span style="display:flex;"><span>        tp<span style="color:#f92672">=</span>model(tu)
</span></span><span style="display:flex;"><span>        loss<span style="color:#f92672">=</span>lossfn(tp,tc)
</span></span><span style="display:flex;"><span>        opt<span style="color:#f92672">.</span>zero_grad()
</span></span><span style="display:flex;"><span>        loss<span style="color:#f92672">.</span>backward()
</span></span><span style="display:flex;"><span>        opt<span style="color:#f92672">.</span>step()
</span></span><span style="display:flex;"><span>        <span style="color:#66d9ef">if</span> i <span style="color:#f92672">%</span> <span style="color:#ae81ff">10</span> <span style="color:#f92672">==</span> <span style="color:#ae81ff">0</span> <span style="color:#f92672">or</span> i <span style="color:#f92672">==</span> <span style="color:#ae81ff">1</span>:
</span></span><span style="display:flex;"><span>            print(<span style="color:#e6db74">&#39;Epoch </span><span style="color:#e6db74">%d</span><span style="color:#e6db74">, Loss </span><span style="color:#e6db74">%f</span><span style="color:#e6db74">&#39;</span> <span style="color:#f92672">%</span> (i, float(loss)))
</span></span><span style="display:flex;"><span>    <span style="color:#66d9ef">return</span> model
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">def</span> <span style="color:#a6e22e">f</span>(x,x2):
</span></span><span style="display:flex;"><span>    <span style="color:#66d9ef">return</span> x<span style="color:#f92672">**</span><span style="color:#ae81ff">2</span><span style="color:#f92672">+</span>x2
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>PATH <span style="color:#f92672">=</span> <span style="color:#e6db74">&#34;cyNN.pt&#34;</span>
</span></span><span style="display:flex;"><span>cyNN<span style="color:#f92672">=</span>t<span style="color:#f92672">.</span>load(PATH)
</span></span><span style="display:flex;"><span>cyNN<span style="color:#f92672">.</span>eval()
</span></span><span style="display:flex;"><span>opt<span style="color:#f92672">=</span>optim<span style="color:#f92672">.</span>Adam(cyNN<span style="color:#f92672">.</span>parameters(),lr<span style="color:#f92672">=</span><span style="color:#ae81ff">1e-7</span>)
</span></span><span style="display:flex;"><span>    <span style="color:#960050;background-color:#1e0010">、、、、、、、、、、</span>
</span></span></code></pre></div><p>我们这个模型拟合的函数是：f(x,x2)=x^2+x2</p>
<h4 id="最最重要的核心调用框架">最最重要的核心调用框架</h4>
<p><img src="/images/1741976905763.jpg" alt=""></p>
<hr>
<h2 id="神经网络解释">神经网络解释</h2>
<p>这是一个四层人工神经网络，拟合二元二次函数。</p>
<p>输入层有两个神经元，输出层有一个神经元，中间两层各有40个神经元，最后一层有24个神经元。激活函数使用的是ReLU，优化器使用的是Adam，学习率为1e-7，损失函数使用的是MSELoss。</p>
<pre tabindex="0"><code>输入层：2个神经元  （表示两个自变量）
隐藏层1：40个神经元
隐藏层2：40个神经元
隐藏层3：24个神经元
输出层：1个神经元   （表示一个因变量）
激活函数：ReLU （线性整流函数）
优化器：Adam 
学习率：1e-7 
损失函数：MSELoss(即均方误差，也就是绝对误差的平方)
</code></pre><h4 id="结构图">结构图</h4>
<p><img src="/images/1741976913419.jpg" alt=""><br>
(一目了然这个属于是=。=）</p>
        <a href="https://cywd123.github.io/posts/%E6%9C%BA%E5%99%A8%E5%AD%A6%E4%B9%A0%E5%9B%9B%E5%B1%82%E4%BA%BA%E5%B7%A5%E7%A5%9E%E7%BB%8F%E7%BD%91%E7%BB%9C%E6%8B%9F%E5%90%88%E4%BA%8C%E5%85%83%E4%BA%8C%E6%AC%A1%E5%87%BD%E6%95%B0/" class="post-read-more">[Read More]</a>
        
    </div>

    
    <div class="blog-tags">
        
       
       <a href="https://cywd123.github.io/tags/%E7%BC%96%E7%A8%8B/">编程</a>&nbsp;
        
       
       <a href="https://cywd123.github.io/tags/%E6%9C%BA%E5%99%A8%E5%AD%A6%E4%B9%A0/">机器学习</a>&nbsp;
         
    </div>
    

</article>
          
            <article class="post-preview">
    <a href="https://cywd123.github.io/posts/%E6%9C%BA%E5%99%A8%E5%AD%A6%E4%B9%A0%E5%8F%AF%E6%8B%9F%E5%90%88%E4%BB%BB%E6%84%8F%E7%BA%BF%E6%80%A7%E5%87%BD%E6%95%B0%E7%9A%84%E5%B0%8F%E7%8E%A9%E6%84%8F%E5%84%BF/">
        <h2 class="post-title">【机器学习】可拟合任意线性函数的小玩意儿</h2>
        
        
        
    </a>

    <p class="post-meta">
        <span class="post-meta">
  
  
  <i class="fas fa-calendar"></i>&nbsp;Posted on March 27, 2023
  
    &nbsp;(Last modified on December 1, 2025)
  
  
  
  
    
      &nbsp;|&nbsp;<i class="fas fa-user"></i>&nbsp;
    
  
  
</span>


    </p>
    <div class="post-entry">
        
        <h1 id="机器学习可拟合任意线性函数的小玩意儿">【机器学习】可拟合任意线性函数的小玩意儿</h1>
<h3 id="先上完整代码">先上完整代码</h3>
<pre tabindex="0"><code>import torch as t
import torch.nn as nn
import torch.optim as optim

def training_loop(n_epochs, optimizer, model,loss_fn, t_u, t_c):
    for epoch in range(1, n_epochs + 1):
        t_p = model(t_u)          #计算当前模型的预测值
        loss = loss_fn(t_p, t_c)  #计算损失函数
        optimizer.zero_grad()   
        loss.backward()   #回溯累计损失
        optimizer.step()
        if epoch % 50 == 0:
            print(&#39;Epoch %d, Loss %f&#39; % (epoch, float(loss)))
    return model


epoches=30000  #迭代次数 一般次数适中，为防止过度拟合不建议迭代次数过多
num=50000  #随机生成的样本数 一般越多越好，看运行配置

learning_rate=1e-4   #学习率（数值越大学得越快（但不一定准确），数值越小拟合越好（时间较长））建议取 1e-2 、1e-3、1e-4、1e-5

x = t.randn(num, 1)
def f(x):
    return 30 * x + 10
    #待拟合的式子为 y=30*x +10  可自行修改if you like。


y = f(x)+ t.randn(num, 1)*10  #  后面的 t.randn(num, 1)*10 是偏离值，降低训练的精确性（给你的模型更多的挑战）
linear_model = nn.Linear(1, 1)
optimizer = optim.SGD(linear_model.parameters(), lr=learning_rate)
training_loop(n_epochs=epoches,optimizer=optimizer,model=linear_model,loss_fn=nn.MSELoss(),t_u=x,t_c=y)
print(&#39;\n&#39;)
#print(linear_model.weight)
#print(linear_model.bias)

while(1):
    x=float(input(&#39;请输入想预测的x值：&#39;))
    tmp=float(linear_model(t.tensor([x]))[0])
    print(&#34;拟合值:&#34;,tmp)
    print(&#34;真实值:&#34;,f(x))
    print(&#34;相对误差:&#34;,1-tmp/f(x))
    print(&#39;\n&#39;)
</code></pre><p>这个程序可以分为两个部分(生成训练数据、机器学习模型)，其中生成训练数据部分可有可无(因为在实际应用中，一般数据为人工导入，不是系统随机生成)。</p>
        <a href="https://cywd123.github.io/posts/%E6%9C%BA%E5%99%A8%E5%AD%A6%E4%B9%A0%E5%8F%AF%E6%8B%9F%E5%90%88%E4%BB%BB%E6%84%8F%E7%BA%BF%E6%80%A7%E5%87%BD%E6%95%B0%E7%9A%84%E5%B0%8F%E7%8E%A9%E6%84%8F%E5%84%BF/" class="post-read-more">[Read More]</a>
        
    </div>

    
    <div class="blog-tags">
        
       
       <a href="https://cywd123.github.io/tags/%E7%BC%96%E7%A8%8B/">编程</a>&nbsp;
        
       
       <a href="https://cywd123.github.io/tags/%E6%9C%BA%E5%99%A8%E5%AD%A6%E4%B9%A0/">机器学习</a>&nbsp;
         
    </div>
    

</article>
          
            <article class="post-preview">
    <a href="https://cywd123.github.io/posts/%E7%88%AC%E8%99%AB%E8%B1%86%E7%93%A3top250%E7%88%AC%E5%8F%96/">
        <h2 class="post-title">【爬虫】豆瓣Top250爬取</h2>
        
        
        
    </a>

    <p class="post-meta">
        <span class="post-meta">
  
  
  <i class="fas fa-calendar"></i>&nbsp;Posted on March 26, 2023
  
    &nbsp;(Last modified on December 1, 2025)
  
  
  
  
    
      &nbsp;|&nbsp;<i class="fas fa-user"></i>&nbsp;
    
  
  
</span>


    </p>
    <div class="post-entry">
        
        <h1 id="豆瓣top250爬取">豆瓣Top250爬取</h1>
<p><img src="/images/1741976760243.jpg" alt=""></p>
<h5 id="这里是简单的代码">这里是简单的代码</h5>
<p><img src="/images/1741976763554.jpg" alt=""></p>
<h5 id="这里是不断修改ua的过程每次请求都会随机生成一个ua这样就可以绕过豆瓣的反爬虫机制了">这里是不断修改UA的过程，每次请求都会随机生成一个UA，这样就可以绕过豆瓣的反爬虫机制了</h5>
<p><img src="/images/1741976779793.jpg" alt=""></p>
<h5 id="这里是爬虫的结果">这里是爬虫的结果</h5>
<p><img src="/images/1741976793865.jpg" alt=""></p>
<h5 id="结果被豆瓣网站检测到了不过我们的爬虫已经圆满的完成了任务我们可以看到豆瓣网站的反爬虫机制是通过检测请求头中的user-agent来判断是否是爬虫所以我们可以通过修改请求头中的user-agent来绕过反爬虫机制">结果被豆瓣网站检测到了，不过我们的爬虫已经圆满的完成了任务，我们可以看到豆瓣网站的反爬虫机制是通过检测请求头中的User-Agent来判断是否是爬虫，所以我们可以通过修改请求头中的User-Agent来绕过反爬虫机制。</h5>
<p><img src="/images/1741976799043.jpg" alt=""></p>

        
    </div>

    
    <div class="blog-tags">
        
       
       <a href="https://cywd123.github.io/tags/%E7%BC%96%E7%A8%8B/">编程</a>&nbsp;
        
       
       <a href="https://cywd123.github.io/tags/%E7%88%AC%E8%99%AB/">爬虫</a>&nbsp;
         
    </div>
    

</article>
          
            <article class="post-preview">
    <a href="https://cywd123.github.io/posts/%E4%B9%B0%E4%B9%A6%E5%AD%A6%E4%B9%A0%E5%96%BD/">
        <h2 class="post-title">买书学习喽</h2>
        
        
        
    </a>

    <p class="post-meta">
        <span class="post-meta">
  
  
  <i class="fas fa-calendar"></i>&nbsp;Posted on March 21, 2023
  
    &nbsp;(Last modified on December 1, 2025)
  
  
  
  
    
      &nbsp;|&nbsp;<i class="fas fa-user"></i>&nbsp;
    
  
  
</span>


    </p>
    <div class="post-entry">
        
        <p>从此时此刻起便是我故事开始的序章！<br>
学起来！</p>

        
    </div>

    

</article>
          
        </div>

        
          <ul class="pager main-pager">
            
              <li class="previous">
                <a href="https://cywd123.github.io/page/4/">&larr; Newer Posts</a>
              </li>
            
            
              <li class="next">
                <a href="https://cywd123.github.io/page/6/">Older Posts &rarr;</a>
              </li>
            
          </ul>
        
      </div>
    </div>
  </div>

      
    <div class="page-meta">
  
  
  
    (Last modified on December 1, 2025)
  
</div>


  <footer>
  <div class="container">
    
    <div class="row">
      <div class="col-lg-8 col-lg-offset-2 col-md-10 col-md-offset-1">
        <ul class="list-inline text-center footer-links">
          
          
        </ul>
        <p class="credits copyright text-muted">
          

          &nbsp;&bull;&nbsp;&copy;
          
            2025
          

          
            &nbsp;&bull;&nbsp;
            <a href="https://cywd123.github.io/">Juiceright</a>
          
        </p>
        
        <p class="credits theme-by text-muted">
          <a href="https://gohugo.io">Hugo v0.152.2</a> powered &nbsp;&bull;&nbsp; Theme <a href="https://github.com/halogenica/beautifulhugo">Beautiful Hugo</a> adapted from <a href="https://deanattali.com/beautiful-jekyll/">Beautiful Jekyll</a>
          
        </p>
      </div>
    </div>
  </div>
</footer><script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.22/dist/katex.min.js" integrity="sha384-cMkvdD8LoxVzGF/RPUKAcvmm49FQ0oxwDF3BGKtDXcEc+T1b2N+teh/OJfpU0jr6" crossorigin="anonymous"></script>
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.16.22/dist/contrib/auto-render.min.js" integrity="sha384-hCXGrW6PitJEwbkoStFjeJxv+fSOOQKOPbJxSfM6G5sWZjAyWhXiTIIAmQqnlLlh" crossorigin="anonymous" onload="renderMathInElement(document.body);"></script>
<script src="https://code.jquery.com/jquery-3.7.0.slim.min.js" integrity="sha384-w5y/xIeYixWvfM+A1cEbmHPURnvyqmVg5eVENruEdDjcyRLUSNej7512JQGspFUr" crossorigin="anonymous"></script>
<script src="https://cdn.jsdelivr.net/npm/bootstrap@3.4.1/dist/js/bootstrap.min.js" integrity="sha384-aJ21OjlMXNL5UyIl/XNwTMqvzeRMZH2w8c5cRVpzpU8Y5bApTppSuUkhZXN0VxHd" crossorigin="anonymous"></script>

<script src="https://cywd123.github.io/js/main.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/photoswipe/4.1.2/photoswipe.min.js" integrity="sha384-QELNnmcmU8IR9ZAykt67vGr9/rZJdHbiWi64V88fCPaOohUlHCqUD/unNN0BXSqy" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/photoswipe/4.1.2/photoswipe-ui-default.min.js" integrity="sha384-m67o7SkQ1ALzKZIFh4CiTA8tmadaujiTa9Vu+nqPSwDOqHrDmxLezTdFln8077+q" crossorigin="anonymous"></script><script src="https://cywd123.github.io/js/load-photoswipe.js"></script>










    
  </body>
</html>

